keras_nlp/models/task.py

```diff
@@ -217,60 +226,129 @@ def from_preset(
-                    )
-                if len(subclasses) > 1:
-                    names = ", ".join(f"`{x.__name__}`" for x in subclasses)
-                    raise ValueError(
-                        f"Ambiguous call to `{cls.__name__}.from_preset()`. "
-                        f"Found multiple possible subclasses {names}. "
-                        "Please call `from_preset` on a subclass directly."
+            task = load_serialized_object(preset, TASK_CONFIG_FILE)
+            if load_weights:
+                task.load_task_weights(get_file(preset, TASK_WEIGHTS_FILE))
+                task.backbone.load_weights(get_file(preset, MODEL_WEIGHTS_FILE))
+        else:
+            # Backbone case.
+            backbone_preset_cls = check_config_class(preset, CONFIG_FILE)
+            if issubclass(backbone_preset_cls, Backbone):
+                if backbone_preset_cls is not cls.backbone_cls:
+                    subclasses = list_subclasses(cls)
+                    subclasses = tuple(
+                        filter(
+                            lambda x: x.backbone_cls == backbone_preset_cls,
+                            subclasses,
+                        )
                     )
-                cls = subclasses[0]
-            # Forward dtype to the backbone.
-            config_overrides = {}
-            if "dtype" in kwargs:
-                config_overrides["dtype"] = kwargs.pop("dtype")
-            backbone = load_from_preset(
-                preset,
-                load_weights=load_weights,
-                config_overrides=config_overrides,
-            )
-            if "preprocessor" in kwargs:
-                preprocessor = kwargs.pop("preprocessor")
-            else:
-                tokenizer = load_from_preset(
+                    if len(subclasses) == 0:
+                        raise ValueError(
+                            f"No registered subclass of `{cls.__name__}` can load "
+                            f"a `{backbone_preset_cls.__name__}`."
+                        )
+                    if len(subclasses) > 1:
+                        names = ", ".join(f"`{x.__name__}`" for x in subclasses)
+                        raise ValueError(
+                            f"Ambiguous call to `{cls.__name__}.from_preset()`. "
+                            f"Found multiple possible subclasses {names}. "
+                            "Please call `from_preset` on a subclass directly."
+                        )
+                    cls = subclasses[0]
+                # Forward dtype to the backbone.
+                config_overrides = {}
+                if "dtype" in kwargs:
+                    config_overrides["dtype"] = kwargs.pop("dtype")
+                backbone = backbone_preset_cls.from_preset(
                     preset,
-                    config_file="tokenizer.json",
+                    load_weights=load_weights,
+                    config_overrides=config_overrides,
```

Comment:
Below this, we can have
```
preprocessor = cls.preprocessor_cls.from_preset(preset)
return cls(backbone=backbone, preprocessor=preprocessor, **kwargs)
```
and then we are done I think. No need for the rest of this function. We can delegate to `preprocessor.from_preset` which has the logic you have below.

Reply:
Right! Preprocessor should have all the necessary validation logic to prevent repetition!
Done!